### ISSUE TITLE
[Cache] Argument Count Error with PHP 7.4.12 and APC provider with preloading enabled

### ISSUE THREAD

**Issue by jrjohnson** [symfony/symfony#39010]
**Symfony version(s) affected**: 5.1.8

**Description**  
Using the cache component with the APC provider in PHP 7.4.11 fails with the error message:
`ArgumentCountError: Too few arguments to function Symfony\Component\Cache\Adapter\AbstractAdapter::throwError(), 2 passed in and exactly 4 expected`

**How to reproduce**  
Create a symfony app with opcache preloading enabled and attempt to access items in the cache with APC as the provider
I created a docker based reproduction of this at https://github.com/jrjohnson/reproduction-symfony-error-apc-opcache the error can be reproduced there by doing:

```bash
composer install
docker-compose build
docker-compose up
```

Then visit http://127.0.0.1:8000/ which will fail with a 500 error and messages logged to the console running docker-compose.

**Additional context**  
I'm not sure this is a symfony issue at it's core, but I was unable to reproduce it otherwise. The issue seems to originate in https://github.com/symfony/symfony/blob/5.x/src/Symfony/Component/Cache/Traits/FilesystemCommonTrait.php#L158 but it isn't necessary to actually cause an error to trigger the bug possibly due to something the preloader is doing.

**Possible Solution**  
Changing the method signature to
```php
public static function throwError($type, $message, $file = '', $line = 0)
```
Appears to fix the problem, but the documentation for [set_error_handler](https://www.php.net/manual/en/function.set-error-handler.php) don't indicate to me that should be necessary so I hesitate to make that change without understanding why this is happening.

**Comment by jderusse** [comment]
Looks similar to #39005 and #38976, can you try disabling the `preload` and check if it fixes the issue?

**Comment by derrabus** [comment]
@jderusse This is a different issue.

@jrjohnson Your suggested fix is correct. Would you like to submit a PR?

**Comment by derrabus** [comment]
4.4 is also affected, so we need to fix the issue on the 4.4 branch.

**Comment by jrjohnson** [comment]
@derrabus, sure, I'll do that right now against 4.4.

**Comment by jderusse** [comment]
Does `set_error_handler` expect a callback with 2 arguments now? Sounds like a BC break in PHP. 

Should we update all occurencies of `set_error_handler`, or just this one?

**Comment by derrabus** [comment]
> but the documentation for set_error_handler don't indicate to me that should be necessary

It actually does!

![Bildschirmfoto 2020-11-05 um 22 05 17](https://user-images.githubusercontent.com/1506493/98296395-27aa2080-1fb3-11eb-8998-c315a52fa134.png)

**Comment by derrabus** [comment]
> Does `set_error_handler` expect a callback with 2 arguments now? Sounds like a BC break in PHP.

The error handler _might_ get called with two arguments only. This is rare, but it might happen. In this case, I would guess that the error is triggered by `ext-apcu` internally, so we don't have a file and line where the error happened.

> Should we update all occurencies of `set_error_handler`, or just this one?

Double checking all custom error handlers might be a good idea.

**Comment by nicolas-grekas** [comment]
I'd suggest opening an issue on the apc repository. This is unexpected behavior.

**Comment by derrabus** [comment]
@jrjohnson The docker compose environment of your reproducer does not build because the `docker/php.ini` file is missing.

**Comment by jrjohnson** [comment]
@nicolas-grekas I'm not sure this about actually throwing an error. If I fix the method signature that exception isn't actually touched so I don't think APC is involved (just the path to the error for me). This also works in PHP 7.4.10 so I think it is something going on with the new opcache preloader. Maybe an optimization because the method is static?

@derrabus sorry, I had that ignored globally, it's been added now.

**Comment by derrabus** [comment]
Thanks, I can reproduce the problem now. Your Dockerfile pulls the latest php 7.4 release which is 7.4.12 actually.

The issue is gone when I do one of the following things:

* downgrade to 7.4.11
* disable preloading
* apply the suggested change

@nicolas-grekas I'm not yet sure what role `ext-apcu` plays in this context. The code that is executed while the error handler is active performs file system operations only, if I'm not mistaken. So my assumption above might be wrong.

To me this looks like a preloading issue, so we should rather report this to php core. However, I expect the answer to be that our error handler does not have the documented signature.

**Comment by derrabus** [comment]
All right, we have quite some error handlers that have similar signatures. While the arguments 3 and 4 are documented as optional, it apparently wasn't a big problem so far that they weren't in our error handlers.

What's also strange is what @jrjohnson reports:

> If I fix the method signature that exception isn't actually touched

This is true. The error handler is not called if we "fix" the signature.

Preloading is really acting strange in 7.4.12.

**Comment by jrjohnson** [comment]
I'm hesitating to make a blanket method signature replacement of every error handler because there are so many that appear to be working now and in some areas like [HttpKernel](https://github.com/symfony/symfony/blob/5.x/src/Symfony/Component/HttpKernel/Kernel.php%23L486) the values for `$file` and `$line` are being passed back into other things. Logically I would guess that providing a default value should do nothing in these cases since the parameter was required in the past, but I worry about munging some behavior that was accidentally being relied on.

I'm happy to do the grunt work of making these changes and I will have a PR done fairly soon, but just wanted to share my concerns.

**Comment by derrabus** [comment]
Fixing every error handler is a really heavy change. Let's not do that unless we're certain that we have to do it.

My suggestion: Only fix the one error handler that's causing trouble for you and leave the rest as they are.

**Comment by jrjohnson** [comment]
Thanks @derrabus, I've opened #39012 to fix it here.

**Comment by nicolas-grekas** [comment]
Clearly that's not an issue with our error handler callbacks: they worked for years.
We would to a disservice to the PHP community if we silently worked around the issue and didn't report this to php.
This is the third report about argument count errors in a few days (#39005 and #38976). Maybe this is a random coincidence, maybe not, but this is very suspicious. We should consider that these might have the same underlying cause.

Please open a PHP bug report.

/cc @nikic in case this rings a bell to you.

**Comment by nikic** [comment]
@nicolas-grekas Yes, I've seen a few reports of preloading issues with PHP 7.4.12. My suspicion is that this issues has been introduced by https://github.com/php/php-src/commit/3928b6b435f6c7c818e8211b2ecdeeb7191310d1 (the only preloading related commit in that release) and that the issue will be fixed by https://github.com/php/php-src/commit/6808968c896a18d2b9f309b2fe63d529058b78b4.

**Comment by nicolas-grekas** [comment]
Thanks for the feedback @nikic, closing then as this is likely not a Symfony issue.

**Comment by PhilETaylor** [comment]
@nikic Ive just rebuilt from PHP8 source this afternoon and now preloading **is completely broken**, worse, php-fpm cannot even start if you attempt to preload :-( 

edit: ok I panicked :) I debugged it down to a missing `use This\Missing\Something` line in a class that was referenced as a type hint on a method param like ` public function doThis(Something $something)` even though that could have been valid and `Something` in the same namespace, it wasn't, and because of that PHP-FPM refused to start... with zero error reporting. Just a tip based on my debugging: sometimes you have to manually start the `php-fpm -O -F` process, and gradually delete classes from the preload file until you get to the one class stopping PHP loading, and then dig deep into that class deleting parts of its structure until the problem was obvious... time lost: 3hours.

**Comment by jrjohnson** [comment]
I can now confirm that this is resolved with PHP 7.4.13 (As well as the latest 8.0RC).
